pkg/apis/core/validation/validation_declarative.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+
+	fldtest "k8s.io/apimachinery/pkg/util/validation/field/testing"
+	validationmetrics "k8s.io/component-base/metrics/prometheus/validation"
+)
+
+// gatherDeclarativeValidationMismatches compares imperative and declarative validation errors
+// and returns detailed information about any mismatches found.  Errors are compared via type, field, and origin
+func gatherDeclarativeValidationMismatches(imperativeErrs, declarativeErrs field.ErrorList) []string {
+	// Only compare imperative errors that are actually CoveredByDeclarative.
+	coveredImperative := make(field.ErrorList, 0, len(imperativeErrs))
+	for _, err := range imperativeErrs {
+		if err.CoveredByDeclarative {
+			coveredImperative = append(coveredImperative, err)
+		}
+	}
+
+	matcher := fldtest.ErrorMatcher{}.ByType().ByField().ByOrigin().RequireOriginWhenInvalid()
+
+	matchedDeclarative := make([]bool, len(declarativeErrs))
+	var mismatchDetails []string
+
+	// Match each "covered" imperative error to a declarative error.
+	for _, iErr := range coveredImperative {
+		foundMatch := false
+
+		for j, dErr := range declarativeErrs {
+			if !matchedDeclarative[j] && matcher.Matches(iErr, dErr) {
+				matchedDeclarative[j] = true
+				foundMatch = true
+				break
+			}
+		}
+
+		if !foundMatch {
+			mismatchDetails = append(mismatchDetails,
+				fmt.Sprintf(
+					"Imperative validation error marked as CoveredByDeclarative but missing from declarative validation: %s",
```

Comment:
Let's phrase this differently.

If the system is working correctly this should never happen, so being extra clear on what went wrong, the severity of it, and what action the user should take is worth including here.

I'm thinking phrasing along the lines of:

"Unexpected difference between hand written validation and declarative validation error results ... "
"this indicates a major bug in the implementation of declarative validation, please disable DeclarativeValidation feature gate to correct this problem" (if applicable)

Reply:
I've updated this now, the text now reads:

```golang
				fmt.Sprintf(
					"Unexpected difference between hand written validation and declarative validation error results, unmatched error(s) found %s. "+
						"This indicates a major bug in the implementation of declarative validation, please disable DeclarativeValidation feature gate to correct this problem",
					matcher.Render(iErr),
				),
```

```golang
				fmt.Sprintf(
					"Unexpected difference between hand written validation and declarative validation error results, extra error(s) found %s. "+
						"This indicates a major bug in the implementation of declarative validation, please disable DeclarativeValidation feature gate to correct this problem",
					matcher.Render(dErr),
				),
```